### ISSUE TITLE
Edge loss didn't converge

### ISSUE THREAD

**Issue by GitJjy** [geovsion/EaNet#1]
Hi @geovsion ,
I just added edge loss (EdgeLoss module) in your code to the original cross entropy loss of a semantic segmentation network (DDCM-Net) , then trained the network on the ISPRS Vaihingen and Potsdam dataset, but the edge loss part didn't converge and kept oscillating like this:
![edge_loss](https://user-images.githubusercontent.com/26278633/104540281-19831b00-565a-11eb-8279-d0fc776b3c30.jpg)
I visualize the edge map in the forward method of EdgeLoss module as follows:
`def forward(self, logits, label):
        prediction = F.softmax(logits, dim=1)
        edge_maps = []
        # kernel size of Laplace filter
        ks = 2 * self.radius
        filt1 = torch.ones(1, 1, ks, ks)
        filt1[:, :, self.radius:2*self.radius, self.radius:2*self.radius] = -8
        filt1.requires_grad = False
        filt1 = filt1.cuda()
        label = label.unsqueeze(1)
        # apply 2D convolution over label using filt1 (Laplacian filter) to get label edge map
        lbedge = F.conv2d(label.float(), filt1, bias=None, stride=1, padding=self.radius)
        lbedge = 1 - torch.eq(lbedge, 0).float()
        
        filt2 = torch.ones(self.n_classes, 1, ks, ks)
        filt2[:, :, self.radius:2*self.radius, self.radius:2*self.radius] = -8
        filt2.requires_grad = False
        filt2 = filt2.cuda()
        # get edge map of prediction
        prededge = F.conv2d(prediction.float(), filt2, bias=None,
                            stride=1, padding=self.radius, groups=self.n_classes)
        
        # squash C-dimensional (C is n_classes) edge map to single edge map
        norm = torch.sum(torch.pow(prededge,2), 1).unsqueeze(1)
        prededge = norm/(norm + self.alpha)
        
        
        # mask = lbedge.float()
        # num_positive = torch.sum((mask==1).float()).float()
        # num_negative = torch.sum((mask==0).float()).float()

        # mask[mask == 1] = 1.0 * num_negative / (num_positive + num_negative)
        # mask[mask == 0] = 1.5 * num_positive / (num_positive + num_negative)
    
        # cost = torch.nn.functional.binary_cross_entropy(
            # prededge.float(),lbedge.float(), weight=mask, reduce=False)
        # return torch.mean(cost)

        
        # visualize edge maps
        edge_maps.append(prededge.float())
        edge_maps.append(lbedge.float())
        edge_maps = torch.cat(edge_maps, dim=0)
        edge_maps_grid = torchvision.utils.make_grid(edge_maps, nrow=prededge.size(0))
        plt.imshow(edge_maps_grid.permute(1, 2, 0).detach().cpu(), cmap='gray')
        plt.title('Edge maps of 1-channel pred and gt')
        plt.show()
        
        return BinaryDiceLoss()(prededge.float(),lbedge.float())`
The visualized result on the Vaihingen dataset is like this:
![edge_map](https://user-images.githubusercontent.com/26278633/104541467-a333e800-565c-11eb-89c1-cce1593fbaa2.png)
The optimizer is Adam with AMSGrad. May you help me find the reason? Thanks!

**Comment by Geo-Tell** [comment]
Sorry for the late response (because the email alert did not work these days).  I have checked the project and found that there is a typo in the code, where 'ks = 2* self.radius' should be 'ks = 2*self.radius +1'. 
I revised the code before updating, but accidentally made somes typos. The code is now renewed. 
Thank you for your question.

**Comment by GitJjy** [comment]
Thanks! @geovsion
